File: motion/utilities/AMCEditor.cs
/*
 * AMCEditor.cs - An editor for AMC files, based loosely on the IPO
 *	interface in blender
 *
 * Copyright (C) 2005-2006 David Trowbridge
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

using System.Collections;

class AMCEditor
{
	[Glade.Widget] Gtk.Window		toplevel;
	[Glade.Widget] Gtk.TreeView		bone_list;
	[Glade.Widget] CurveEditor		curve_editor;
	[Glade.Widget] Gtk.EventBox		eventbox;
	[Glade.Widget] Gtk.ImageMenuItem	menu_save;
	[Glade.Widget] Gtk.ImageMenuItem	menu_saveas;

	Gtk.TreeStore				bone_store;
	AMC.File				AMCData;
	string					Filename;

	bool					modified;

	public static void
	Main (string[] args)
	{
		Gtk.Application.Init ();
		new AMCEditor ();
		Gtk.Application.Run ();
	}

	private
	AMCEditor ()
	{
		Glade.XML.SetCustomHandler (new Glade.XMLCustomWidgetHandler (GladeCustomHandler));
		Glade.XML gxml = new Glade.XML (null, "amc-editor.glade", "toplevel", null);
		gxml.Autoconnect (this);

		// Create the tree store
		bone_store = new Gtk.TreeStore (
				typeof (string),	// name
				typeof (Gdk.Color),	// color
				typeof (bool),		// shown on curve window
				typeof (bool),		// whether toggle is visible
				typeof (int),		// data index
				typeof (int));		// color index
		bone_list.Model = bone_store;

		// Create our text column
		Gtk.TreeViewColumn text_column  = new Gtk.TreeViewColumn ();
		Gtk.CellRendererText text_renderer = new Gtk.CellRendererText ();
		text_column.PackStart (text_renderer, true);
		text_column.AddAttribute (text_renderer, "markup",         0);
		text_column.AddAttribute (text_renderer, "foreground-gdk", 1);
		bone_list.AppendColumn (text_column);

		// Create our visible column
		Gtk.TreeViewColumn visible_column = new Gtk.TreeViewColumn ();
		Gtk.CellRendererToggle visible_renderer = new Gtk.CellRendererToggle ();
		visible_column.PackStart (visible_renderer, false);
		visible_column.AddAttribute (visible_renderer, "active",  2);
		visible_column.AddAttribute (visible_renderer, "visible", 3);
		visible_renderer.Activatable = true;
		visible_renderer.Toggled += new Gtk.ToggledHandler (RowToggled);
		bone_list.AppendColumn (visible_column);

		Filename = null;
		modified = false;

		curve_editor.bone_store = bone_store;

		UpdateToolbarSensitivity ();

		toplevel.ShowAll ();
	}

	void
	RowToggled (object o, Gtk.ToggledArgs args)
	{
		Gtk.TreeIter iter;
		bone_store.GetIter (out iter, new Gtk.TreePath (args.Path));
		bool t = (bool) bone_store.GetValue (iter, 2);
		bone_store.SetValue (iter, 2, !t);

		if (!t)
			curve_editor.enabled_bones.Add (args.Path);
		else
			curve_editor.enabled_bones.Remove (args.Path);

		curve_editor.VisibilityChanged ();
		eventbox.GrabFocus ();
	}

	static Gtk.Widget
	GladeCustomHandler (Glade.XML xml, string func_name, string name, string str1, string str2, int int1, int int2)
	{
		if (func_name == "CurveEditor")
			return new CurveEditor ();
		return null;
	}

	// Signal handlers
	public void
	OnOpen (object o, System.EventArgs args)
	{
		object[] responses = {
			Gtk.Stock.Cancel, Gtk.ResponseType.Cancel,
			Gtk.Stock.Open,   Gtk.ResponseType.Accept,
		};
		Gtk.FileChooserDialog fs = new Gtk.FileChooserDialog ("Open AMC...", null, Gtk.FileChooserAction.Open, responses);

		// File filter - *.amc
		Gtk.FileFilter ff1 = new Gtk.FileFilter ();
		ff1.AddPattern ("*.amc");
		ff1.Name = "Acclaim Motion Capture (*.amc)";
		fs.AddFilter (ff1);

		// File filter - all files
		Gtk.FileFilter ff2 = new Gtk.FileFilter ();
		ff2.AddPattern ("*");
		ff2.Name = "All Files";
		fs.AddFilter (ff2);

		Gtk.ResponseType response = (Gtk.ResponseType) fs.Run ();
		fs.Hide ();

		if (response == Gtk.ResponseType.Accept) {
			Filename = fs.Filename;
			AMCData = AMC.File.Load (Filename);

			bone_store.Clear ();

			if (AMCData == null) {
				// FIXME - pop up an error dialog
				System.Console.WriteLine ("Error loading {0}", Filename);
				Filename = null;
			} else {
				curve_editor.AMC = AMCData;
				SetTitle ();

				AMC.Frame f = (AMC.Frame) AMCData.frames[0];

				IDictionaryEnumerator e = f.GetEnumerator ();
				e.Reset ();
				while (e.MoveNext ()) {
					Gtk.TreeIter iter;
					iter = bone_store.AppendNode ();
					bone_store.SetValue (iter, 0, e.Key);
					bone_store.SetValue (iter, 1, new Gdk.Color (0x00, 0x00, 0x00));

					float[] s = (float[]) e.Value;
					for (int i = 0; i < s.Length; i++) {
						int color = ColorMap.AllocColor ();
						Gtk.TreeIter citer = bone_store.AppendNode (iter);
						bone_store.SetValue (citer, 0, System.String.Format ("<b>{0}</b>", i.ToString ()));
						bone_store.SetValue (citer, 1, ColorMap.Colors[color]);
						bone_store.SetValue (citer, 3, true);
						bone_store.SetValue (citer, 4, color);
						bone_store.SetValue (citer, 5, i);
					}
				}

				modified = false;
				UpdateToolbarSensitivity ();
			}
		}
		fs.Destroy ();
	}

	public void
	OnSave (object o, System.EventArgs args)
	{
		if (Filename == null) {
			object[] responses = {
				Gtk.Stock.Cancel, Gtk.ResponseType.Cancel,
				Gtk.Stock.Save,   Gtk.ResponseType.Accept,
			};

			Gtk.FileChooserDialog fs = new Gtk.FileChooserDialog ("Save...", null, Gtk.FileChooserAction.Save, responses);
			fs.DefaultResponse = Gtk.ResponseType.Accept;
			fs.CurrentName = Filename;

			Gtk.ResponseType response = (Gtk.ResponseType) fs.Run ();

			if (response == Gtk.ResponseType.Accept) {
				Filename = fs.Filename;
				AMCData.Save (Filename);
				UpdateToolbarSensitivity ();
			}
			fs.Destroy ();
		} else {
			AMCData.Save (Filename);
			modified = false;
		}
	}

	public void
	OnSaveAs (object o, System.EventArgs args)
	{
		object[] responses = {
			Gtk.Stock.Cancel, Gtk.ResponseType.Cancel,
			Gtk.Stock.Save,   Gtk.ResponseType.Accept,
		};

		Gtk.FileChooserDialog fs = new Gtk.FileChooserDialog ("Save As...", null, Gtk.FileChooserAction.Save, responses);
		fs.DefaultResponse = Gtk.ResponseType.Accept;
		fs.CurrentName = Filename;

		Gtk.ResponseType response = (Gtk.ResponseType) fs.Run ();

		if (response == Gtk.ResponseType.Accept) {
			Filename = fs.Filename;
			AMCData.Save (Filename);
			UpdateToolbarSensitivity ();
		}
		fs.Destroy ();
	}

	public void
	OnQuit (object o, System.EventArgs args)
	{
		Gtk.Application.Quit ();
	}

	void
	SetTitle ()
	{
		if (Filename == null) {
			toplevel.Title = "AMC Editor";
		} else {
			if (modified)
				toplevel.Title = System.IO.Path.GetFileName (Filename) + "*";
			else
				toplevel.Title = System.IO.Path.GetFileName (Filename);
		}
	}

	void
	UpdateToolbarSensitivity ()
	{
		menu_save.Sensitive = modified;
		menu_saveas.Sensitive = (Filename != null);
	}

	public void
	CurveButtonPressEvent (object o, Gtk.ButtonPressEventArgs args)
	{
		eventbox.GrabFocus ();
		curve_editor.ButtonPress (args.Event);
	}

	public void
	CurveButtonReleaseEvent (object o, Gtk.ButtonReleaseEventArgs args)
	{
		curve_editor.ButtonRelease (args.Event);
	}

	public void
	CurveMotionNotifyEvent (object o, Gtk.MotionNotifyEventArgs args)
	{
		curve_editor.MotionNotify (args.Event);
	}

	public void
	CurveKeyPressEvent (object o, Gtk.KeyPressEventArgs args)
	{
		curve_editor.KeyPress (args.Event);
	}
}

File: py/amc_editor.py
#!/usr/bin/env python
#
# amc_editor.py - An editor for AMC files, based loosely on the IPO
#       interface in blender
#
# Copyright (C) 2005-2006 David Trowbridge
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#

import os
import sys

import pygtk
pygtk.require ('2.0')
import gobject
import gtk
import gtk.glade

from Motion import AMC
from amc_editor import Color
from amc_editor.CurveEditor import CurveEditor

class AMCEditor:
    def __init__ (self):
        self.widgets = {}
        self.modified = False
        self.filename = None

        gtk.glade.set_custom_handler (self.glade_custom_handler)

        gladefile = os.path.join (os.path.dirname(sys.argv[0]), 'amc_editor', 'amc-editor.glade')
        self._glade = gtk.glade.XML (gladefile)
        self._glade.signal_autoconnect (self)

        def get_widget (name):
            self.widgets[name] = self._glade.get_widget (name)
        get_widget ('bone_list')
        get_widget ('curve_editor')
        get_widget ('eventbox')
        get_widget ('menu_save')
        get_widget ('menu_saveas')
        get_widget ('toplevel')
        get_widget ('hscroll')
        get_widget ('vscroll')

        self._bone_store = gtk.TreeStore (str,           # name
                                          gtk.gdk.Color, # color
                                          'gboolean',    # shown on curve window
                                          'gboolean',    # whether toggle is visible
                                          str,           # bone index
                                          int,           # dof index
                                          int            # color index
                                          )
        self.widgets['bone_list'].set_model (self._bone_store)

        self.widgets['curve_editor'].set_scroll_adjustments (self.widgets['hscroll'].get_adjustment (),
                                                             self.widgets['vscroll'].get_adjustment ())
        self.widgets['curve_editor']._bone_store = self._bone_store

        # Create text column
        text_column   = gtk.TreeViewColumn ()
        text_renderer = gtk.CellRendererText ()
        text_column.pack_start (text_renderer, True)
        text_column.add_attribute (text_renderer, 'markup'        , 0)
        text_column.add_attribute (text_renderer, 'foreground-gdk', 1)
        self.widgets['bone_list'].append_column (text_column)

        # Create visibility toggle column
        visible_column   = gtk.TreeViewColumn ()
        visible_renderer = gtk.CellRendererToggle ()
        visible_column.pack_start (visible_renderer, False)
        visible_column.add_attribute (visible_renderer, 'active',  2)
        visible_column.add_attribute (visible_renderer, 'visible', 3)
        visible_renderer.activatable = True
        visible_renderer.connect ('toggled', self.row_toggled)
        self.widgets['bone_list'].append_column (visible_column)

        self.update_toolbar_sensitivity ()
        self.set_title ()
        self.widgets['toplevel'].show_all ()

    def main (self):
        gtk.main ()

    def row_toggled (self, cell, path):
        iter = self._bone_store.get_iter (path)
        t = not self._bone_store.get_value (iter, 2)
        self._bone_store.set_value (iter, 2, t)

        if (t):
            self.widgets['curve_editor'].enable_bone (path)
        else:
            self.widgets['curve_editor'].disable_bone (path)

    def set_title (self):
        if self.filename is None:
            self.widgets['toplevel'].set_title ('AMC Editor')
            return

        if self.modified:
            self.widgets['toplevel'].set_title (os.path.basename (self.filename) + '*')
        else:
            self.widgets['toplevel'].set_title (os.path.basename (self.filename))

    def update_toolbar_sensitivity (self):
        self.widgets['menu_save'].set_sensitive   (self.modified)
        self.widgets['menu_saveas'].set_sensitive (self.filename is not None)

    # signal handlers
    def on_quit (self, data=None):
        gtk.main_quit ()

    def on_delete_event (self, widget, data=None):
        return False

    def on_open (self, data=None):
        chooser = gtk.FileChooserDialog ('Open...', action = gtk.FILE_CHOOSER_ACTION_OPEN,
                                         buttons = (gtk.STOCK_CANCEL, gtk.RESPONSE_CANCEL,
                                                    gtk.STOCK_OPEN,   gtk.RESPONSE_OK))
        # File filter - *.amc
        filter = gtk.FileFilter ()
        filter.add_pattern ('*.amc')
        filter.set_name ('Acclaim Motion Capture (*.amc)')
        chooser.add_filter (filter)

        # File filter - all files
        filter = gtk.FileFilter ()
        filter.add_pattern ('*')
        filter.set_name ('All files')
        chooser.add_filter (filter)

        response = chooser.run ()
        chooser.hide ()

        if response == gtk.RESPONSE_OK:
            self.filename = chooser.get_filename ()
            amc = AMC.from_file (self.filename)
            self._bone_store.clear ()

            if amc is None:
                # FIXME - pop up error dialog
                print 'Error loading',self.filename
                self.filename = None
            else:
                self.set_title ()

                for (bone, data) in amc.bones.iteritems ():
                    iter = self._bone_store.append (parent = None)
                    self._bone_store.set_value (iter, 0, bone)
                    self._bone_store.set_value (iter, 1, gtk.gdk.Color (0x00, 0x00, 0x00));
                    self._bone_store.set_value (iter, 3, 0)

                    for degree in range (data.shape[1]):
                        iter2 = self._bone_store.append (parent = iter)
                        color = Color.alloc_color ()
                        self._bone_store.set (iter2, 0, '<b>%d</b>' % degree,
                                                     1, Color.colors[color],
                                                     2, 0,
                                                     3, 1,
                                                     4, bone,
                                                     5, degree,
                                                     6, color,
                                                     )
                self.widgets['curve_editor'].set_amc (amc)
                self.modified = False
        chooser.destroy ()

    def on_save (self, data=None):
        if self.filename is None:
            chooser = gtk.FileChooserDialog ('Save...', action = gtk.FILE_CHOOSER_ACTION_SAVE,
                                             buttons = (gtk.STOCK_CANCEL, gtk.RESPONSE_CANCEL,
                                                        gtk.STOCK_SAVE,   gtk.RESPONSE_OK))
            chooser.set_default_response (gtk.RESPONSE_OK)
            chooser.set_current_name (self.filename)

            response = chooser.run ()

            if response == gtk.RESPONSE_OK:
                self.filename = chooser.get_filename ()
                self.amc.save (self.filename)
                self.update_toolbar_sensitivity ()

            chooser.destroy ()
        else:
            self.amc.save (self.filename)
            self.modified = False

    def glade_custom_handler (self, glade, func_name, name, str1, str2, int1, int2):
        if func_name == 'CurveEditor':
            return CurveEditor ()

if __name__ == '__main__':
    editor = AMCEditor ()
    editor.main ()
